tests/cpp/test_combined_inner_outer_reduction.cpp

```diff
@@ -34,255 +34,129 @@ namespace nvfuser {
-        mean,
-        rstd,
-        weight,
-        bias,
-        {true, true, true});
-
-    if (dtype == DataType::Half) {
-      layer_norm_results.grad_input =
-          castOp(dtype, layer_norm_results.grad_input);
-      layer_norm_results.grad_bias =
-          castOp(dtype, layer_norm_results.grad_bias);
-      layer_norm_results.grad_weight =
-          castOp(dtype, layer_norm_results.grad_weight);
-    }
-
-    fusion.addOutput(layer_norm_results.grad_input);
-    fusion.addOutput(layer_norm_results.grad_weight);
-    fusion.addOutput(layer_norm_results.grad_bias);
-
-    auto maybe_fp16_options = at::TensorOptions()
-                                  .dtype(data_type_to_aten(dtype))
-                                  .device(at::kCUDA, 0);
-
-    at::Tensor aten_grad_out = at::randn(input_shape, maybe_fp16_options);
-    at::Tensor aten_input = at::randn(input_shape, maybe_fp16_options);
-    at::Tensor aten_weight = at::randn(norm_shape, maybe_fp16_options);
-    at::Tensor aten_bias = at::randn(norm_shape, maybe_fp16_options);
-
-    auto at_weight = c10::optional<at::Tensor>(aten_weight);
-    auto at_bias = c10::optional<at::Tensor>(aten_bias);
-
-    const float kEps = 1e-5;
-    auto aten_results =
-        at::native_layer_norm(aten_input, norm_shape, at_weight, at_bias, kEps);
-    auto aten_output = std::get<0>(aten_results);
-    auto aten_mean = std::get<1>(aten_results);
-    auto aten_rstd = std::get<2>(aten_results);
-
-    FusionExecutorCache fec(std::move(fusion_ptr));
-    std::vector<c10::IValue> aten_inputs = {
-        aten_grad_out,
-        aten_input,
-        aten_mean,
-        aten_rstd,
-        aten_weight,
-        aten_bias};
-    auto cg_outputs = fec.runFusionWithInputs(aten_inputs);
-
-    auto aten_gradients = at::native_layer_norm_backward(
-        aten_grad_out,
-        aten_input,
-        norm_shape,
-        aten_mean,
-        aten_rstd,
-        c10::optional<at::Tensor>(aten_weight),
-        c10::optional<at::Tensor>(aten_bias),
-        {true, true, true});
+// tuple of data type, batch size (outer dim), hidden size (inner dim)
+using CombinedSchedulerParams = std::tuple<DataType, int64_t, int64_t>;
+using CombinedSchedulerTest = NVFuserFixtureParamTest<CombinedSchedulerParams>;
+TEST_P(CombinedSchedulerTest, LayerNormBackward) {
+  std::unique_ptr<Fusion> fusion_ptr = std::make_unique<Fusion>();
+  Fusion& fusion = *fusion_ptr.get();
+  FusionGuard fg(&fusion);
+  auto [dtype, batch_size, hidden_size] = GetParam();
+
+  std::vector<int64_t> norm_shape{hidden_size};
+  std::vector<int64_t> input_shape{batch_size, hidden_size};
+  std::vector<int64_t> outer_shape{batch_size, 1};
+  bool fp16_or_bf16 = dtype == DataType::Half || dtype == DataType::BFloat16;
+  auto grad_out = makeContigTensor(input_shape.size(), dtype);
+  auto input = makeContigTensor(input_shape.size(), dtype);
+  auto mean =
+      makeConcreteTensor(outer_shape, fp16_or_bf16 ? DataType::Float : dtype);
```

Comment:
`mean` and `rstd` can't be bfloat? Why not?

Reply:
if other inputs are `fp16_or_bf16`, `mean` and `rstd` are `float`, otherwise they stay the same as other inputs, e.g. when dtype is `double`.

Reply:
They can be any data type but in the current implementation of layer norm, they are saved as `float` if inputs are `half or bfloat`

Reply:
Got it. I think the code in this function can be simplified by introducing a helper:
```
// Returns float if the input data type is lower than that; otherwise, returns the input data type as is. 
DataType floatOrHigher(DataType);
```

Then you can say 

```
mean = makeConcreteTensor(..., floatOrHigher(dtype));
```
and
```
auto cast_to_float_or_higher = [](TensorView* in) -> TensorView* {
  return maybeCastOp(floatOrHigher(in->dtype()), in);
};
grad_out = cast_to_float_or_higher(grad_out);
...
```

Reply:
Good point. I believe we have many similar cases in our test suites. will check and simplify them together in a separate PR.